chrome/content/zotero/xpcom/itemPaneManager.js

```diff
@@ -0,0 +1,335 @@
+ * @typedef SectionHookArgs
+ * @type {SectionBasicHookArgs & SectionUIHookArgs}
+ * @typedef  {SectionHookArgs & { refresh: () => Promise<void> }} SectionInitHookArgs
+ * A `refresh` is exposed to plugins to allows plugins to refresh the section when necessary,
+ * e.g. item modify notifier callback. Note that calling `refresh` during initialization
+ * have no effect.
+ * @typedef  {SectionHookArgs & { incomingData: { type: "item" | "mode" | "inTrash" | "tabType", value: any } }} SectionDataChangeHookArgs
+ * @typedef  {SectionHookArgs & { event: Event }} SectionEventHookArgs
+ * @typedef ItemDetailsSectionOptions
+ * @type {object}
+ * @property {string} paneID - Unique pane ID
+ * @property {string} pluginID - Set plugin ID to auto remove section when plugin is disabled/removed
+ * @property {SectionL10n & SectionIcon} head - Header options. Icon should be 16*16 and `label` need to be localized
+ * @property {SectionL10n & SectionIcon} sidenav - Sidenav options. Icon should be 20*20 and `tooltiptext` need to be localized
+ * @property {string} [fragment] - Pane fragment as string
+ * @property {(options: SectionInitHookArgs) => void} [onInit]
+ * Lifecycle hook called when section is initialized
+ *
+ * Do:
+ * 1. Initialize data if necessary
+ * 2. Setup hooks, e.g. notifier callback
+ *
+ * NOT do:
+ * 1. Render/refresh UI
+ * @property {(options: SectionBasicHookArgs) => void} [onDestroy]
+ * Lifecycle hook called when section is destroyed
+ *
+ * Do:
+ * 1. Remove data and release resource
+ * 2. Remove hooks, e.g. notifier callback
+ *
+ * NOT do:
+ * 1. Render/refresh UI
+ * @property {(options: SectionDataChangeHookArgs) => boolean} [onDataChange]
+ * Lifecycle hook called when section incoming data change received
+ *
+ * Return `false` to abort the data change, otherwise the data change will be applied
+ * (stored internally and can be accessed by `options.getData` in other hooks).
+ *
+ * Do:
+ * 1. Validate and decide whether to accept the incoming data change
+ * 2. Update the section enabled state with `options.setEnabled`. For example, if the section
+ *   is only enabled in the readers, you can use:
+ * ```js
+ * onDataChange(options) {
+ *   let { incomingData, setEnabled } = options.;
+ *   if (incomingData.type === "tabType") {
+ *     setEnabled(incomingData.value === "reader");
+ *   }
+ *   return true;
+ * }
+ * ```
+ * options.setEnabled(options.tabType === "reader");
+ *
+ * NOT do:
+ * 1. Render/refresh UI
+ * @property {(options: SectionHookArgs) => void | Promise<void>} onRender
+ * Lifecycle hook called when section should do primary render
+ *
+ * Create elements and append them to `options.body`.
+ * - If `onRender` is sync, it is always called when the item pane is rendered;
+ * - If `onRender` is async, it will be called when scrolling into view.
+ *
+ * If the rendering can be slow, you should either make it async, or move async part to `onSecondaryRender`.
```

Comment:
Is there a difference? How would you choose one over the other?

Reply:
Oh, I see a case for secondary render is given below, but I think it's still a little unclear what we expect to be done in synchronous onRender, async onRender, synchronous onSecondaryRender, and async onSecondaryRender.

Reply:
I think this illustrates why/how different `render` works:

<img width="916" alt="image" src="https://github.com/zotero/zotero/assets/33902321/72f34185-bc5e-4de8-86cc-7081de2ad233">

To respond to the user's selecting item as soon as possible, we want the initial rendering to be very fast, thus not allowing any async rendering to happen there.

The differences:
- sync `onRender` is always executed during initial render;
- async `onRender` and any `onSecondaryRender ` are executed when scrolling into view (on need).

Reply:
So you might use an async `onRender` if you only had a single render to do but it was going to be slow? Couldn't you just have a no-op (or maybe a height calculation) for `onRender` in that case and use `onSecondaryRender`? It seems a little confusing to have two different ways of rendering stuff at the same time. What if we just had sync `onRender` and `onSecondaryRender` (or even called them `onRender` and `onAsyncRender`)?

Alternatively, are we sure we want to allow plugins to do sync rendering at all? If these will often be below the fold, why provide a mechanism for rendering ahead of time at all and burning cycles on every item select? What if we only allowed a way to return a height, and then an async render method?

Reply:
> What if we just had sync onRender and onSecondaryRender (or even called them onRender and onAsyncRender

Good point. That makes more sense!

> What if we only allowed a way to return a height, and then an async render method?

Might it be awkward to ask plugins to compute a height, since we have dynamic font size and compact/comfortable, which might contribute to computing the height? It might just be not practical for plugins to compute an exact height of things.

I think in most cases the `onRender` burden would not be very heavy - might be just loading some data and setting attributes to several elements. We could add a timer and throw a warning if it's slow to indicate the plugin is doing something bad here.

Reply:
> Might it be awkward to ask plugins to compute a height, since we have dynamic font size and compact/comfortable, which might contribute to computing the height? It might just be not practical for plugins to compute an exact height of things.

Well, I was thinking it could be optional, but I guess the point is that we do want heights as much as possible to avoid weird scroll behavior, and in the default, non-slow case, computing a height (in canvas? I'm not even sure how) could actually result in as much work, which might then just need to be repeated for an actual render. So then by not allowing for a sync render we'd just making the default case worse…

Reply:
Yeah. So let's try the `onRender` and `onAsyncRender`? Ideally, it would be clear enough.

Reply:
Sounds good.